profiling-ffi/src/crashtracker/datatypes.rs

```diff
@@ -0,0 +1,282 @@
+            inner: Box::into_raw(Box::new(crash_info)),
+        }
+    }
+
+    pub(super) fn take(&mut self) -> Option<Box<datadog_crashtracker::CrashInfo>> {
+        // Leaving a null will help with double-free issues that can
+        // arise in C. Of course, it's best to never get there in the
+        // first place!
+        let raw = std::mem::replace(&mut self.inner, std::ptr::null_mut());
+
+        if raw.is_null() {
+            None
+        } else {
+            Some(unsafe { Box::from_raw(raw) })
+        }
+    }
+}
+
+impl Drop for CrashInfo {
+    fn drop(&mut self) {
+        drop(self.take())
+    }
+}
+
+pub(crate) unsafe fn crashinfo_ptr_to_inner<'a>(
+    crashinfo_ptr: *mut CrashInfo,
+) -> anyhow::Result<&'a mut datadog_crashtracker::CrashInfo> {
+    match crashinfo_ptr.as_mut() {
+        None => anyhow::bail!("crashinfo pointer was null"),
+        Some(inner_ptr) => match inner_ptr.inner.as_mut() {
+            Some(crashinfo) => Ok(crashinfo),
+            None => anyhow::bail!("crashinfo's inner pointer was null (indicates use-after-free)"),
+        },
+    }
+}
+
+/// Returned by [ddog_prof_Profile_new].
+#[repr(C)]
+pub enum CrashInfoNewResult {
+    Ok(CrashInfo),
+    #[allow(dead_code)]
+    Err(Error),
+}
+
+#[repr(C)]
+pub enum CrashtrackerGetCountersResult {
+    Ok([i64; ProfilingOpTypes::SIZE as usize]),
+    #[allow(dead_code)]
+    Err(Error),
+}
+
+impl From<anyhow::Result<[i64; ProfilingOpTypes::SIZE as usize]>>
+    for CrashtrackerGetCountersResult
+{
+    fn from(value: anyhow::Result<[i64; ProfilingOpTypes::SIZE as usize]>) -> Self {
+        match value {
+            Ok(x) => Self::Ok(x),
+            Err(err) => Self::Err(err.into()),
+        }
+    }
+}
+
+/// A generic result type for when a crashtracking operation may fail,
+/// but there's nothing to return in the case of success.
+#[repr(C)]
+pub enum CrashtrackerResult {
+    Ok(
+        /// Do not use the value of Ok. This value only exists to overcome
+        /// Rust -> C code generation.
+        bool,
+    ),
+    Err(Error),
+}
+
+impl From<anyhow::Result<()>> for CrashtrackerResult {
+    fn from(value: anyhow::Result<()>) -> Self {
+        match value {
+            Ok(_) => Self::Ok(true),
+            Err(err) => Self::Err(err.into()),
+        }
+    }
+}
+
+#[repr(C)]
+pub struct StackFrameNames<'a> {
+    colno: Option<u32>,
+    filename: CharSlice<'a>,
+    lineno: Option<u32>,
+    name: CharSlice<'a>,
+}
+
+impl<'a> TryFrom<StackFrameNames<'a>> for datadog_crashtracker::StackFrameNames {
+    type Error = anyhow::Error;
+
+    fn try_from(value: StackFrameNames<'a>) -> Result<Self, Self::Error> {
+        let colno = value.colno;
+        let filename = option_from_char_slice(value.filename)?;
+        let lineno = value.lineno;
+        let name = option_from_char_slice(value.name)?;
+        Ok(Self {
+            colno,
+            filename,
+            lineno,
+            name,
+        })
```

Comment:
sorry, couldn't help myself 😅 
```suggestion
        Self::try_from(&value);
```

Reply:
Thanks, done